// @tanstack/start-client-core/dist/cjs/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const headers = require("./headers.cjs");
const serializer = require("./serializer.cjs");
const ssrClient = require("./ssr-client.cjs");
const createIsomorphicFn = require("./createIsomorphicFn.cjs");
const envOnly = require("./envOnly.cjs");
const createServerFn = require("./createServerFn.cjs");
const json = require("./json.cjs");
const createMiddleware = require("./createMiddleware.cjs");
const registerGlobalMiddleware = require("./registerGlobalMiddleware.cjs");
exports.mergeHeaders = headers.mergeHeaders;
exports.startSerializer = serializer.startSerializer;
exports.hydrate = ssrClient.hydrate;
exports.createIsomorphicFn = createIsomorphicFn.createIsomorphicFn;
exports.clientOnly = envOnly.clientOnly;
exports.serverOnly = envOnly.serverOnly;
exports.applyMiddleware = createServerFn.applyMiddleware;
exports.createServerFn = createServerFn.createServerFn;
exports.execValidator = createServerFn.execValidator;
exports.extractFormDataContext = createServerFn.extractFormDataContext;
exports.flattenMiddlewares = createServerFn.flattenMiddlewares;
exports.serverFnBaseToMiddleware = createServerFn.serverFnBaseToMiddleware;
Object.defineProperty(exports, "serverFnStaticCache", {
  enumerable: true,
  get: () => createServerFn.serverFnStaticCache
});
exports.json = json.json;
exports.createMiddleware = createMiddleware.createMiddleware;
exports.globalMiddleware = registerGlobalMiddleware.globalMiddleware;
exports.registerGlobalMiddleware = registerGlobalMiddleware.registerGlobalMiddleware;
